File: app/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement {
    private static final String[] DIRECTION_IMAGE_PATH = {
            "up.png",
            "down.png",
            "left.png",
            "right.png"
    };
    private static final String[] DIRECTION_STRING = {
            "^",
            "∨",
            "<",
            ">"
    };
    private MapDirection orientation;
    private Vector2d position;

    static final private Vector2d defaultPostion = new Vector2d(2, 2);

    public Animal() {
        this(defaultPostion);
    }

    public Animal(Vector2d position) {
        this.position = position;
        this.orientation = MapDirection.NORTH;
    }

    public MapDirection getOrientation() {
        return orientation;
    }

    @Override
    public Vector2d getPosition() {
        return position;
    }

    @Override
    public String getImageFileName() {
        return DIRECTION_IMAGE_PATH[orientation.ordinal()];
    }

    @Override
    public String toString() {
        return DIRECTION_STRING[orientation.ordinal()];
    }

    public boolean isAt(Vector2d position) {
        return this.position.equals(position);
    }

    public void move(MoveDirection direction, MoveValidator moveValidator) {
        switch (direction) {
            case FORWARD: {
                Vector2d potentialPosition = position.add(orientation.toUnitVector());
                if (moveValidator.canMoveTo(potentialPosition))
                    position = potentialPosition;
                break;
            }
            case BACKWARD: {
                Vector2d potentialPosition = position.subtract(orientation.toUnitVector());
                if (moveValidator.canMoveTo(potentialPosition))
                    position = potentialPosition;
                break;
            }
            case LEFT:
                orientation = orientation.previous();
                break;
            case RIGHT:
                orientation = orientation.next();
                break;
        }
    }

}

File: kotlin/app/src/main/kotlin/agh/ics/oop/model/Animal.kt
package agh.ics.oop.model

class Animal(private var position: Vector2d = Vector2d(2, 2))  {
    private var orientation = MapDirection.NORTH

    fun getPosition(): Vector2d = position

    fun getOrientation(): MapDirection = orientation

    override fun toString(): String = when (orientation) {
            MapDirection.NORTH -> "^"
            MapDirection.EAST -> ">"
            MapDirection.SOUTH -> "v"
            MapDirection.WEST -> "<"
    }

    fun itAt(position: Vector2d): Boolean = this.position == position

    fun move(direction : MoveDirection, worldMap : WorldMap) {
        when (direction) {
            MoveDirection.FORWARD -> {
                val potentialPosition = position + orientation.toUnitVector()
                if (worldMap.canMoveTo(potentialPosition))
                    position = potentialPosition
            }
            MoveDirection.BACKWARD -> {
                val potentialPosition = position - orientation.toUnitVector()
                if (worldMap.canMoveTo(potentialPosition))
                    position = potentialPosition
            }
            MoveDirection.LEFT -> orientation = orientation.previous()
            MoveDirection.RIGHT -> orientation = orientation.next()
        }
    }

    fun setPosition(position: Vector2d) {
        this.position = position
    }
 
}
